java/client/src/test/java/glide/ExceptionHandlingTests.java

```diff
@@ -0,0 +1,310 @@
+        var response = Response.newBuilder().setClosingError("TEST").build();
+        callbackDispatcher.completeRequest(response);
+
+        var exception = assertThrows(ExecutionException.class, future1::get);
+        // a ClosingException thrown from CallbackDispatcher::completeRequest and then
+        // rethrown by CommandManager::exceptionHandler
+        assertTrue(exception.getCause() instanceof ClosingException);
+        // check the channel
+        assertTrue(channelHandler.wasClosed);
+
+        // all pending requests should be aborted once ClosingError received in callback dispatcher
+        exception = assertThrows(ExecutionException.class, future2::get);
+        // or could be cancelled in CallbackDispatcher::shutdownGracefully
+        // cancellation overwrites previous status, so we may not get ClosingException due to a race
+        assertTrue(
+                exception.getCause() instanceof ClosingException
+                        || exception.getCause() instanceof CancellationException);
+    }
+
+    private static Stream<Arguments> getProtobufErrorsToJavaClientErrorsMapping() {
+        return Stream.of(
+                Arguments.of(Unspecified, RequestException.class),
+                Arguments.of(ExecAbort, ExecAbortException.class),
+                Arguments.of(Timeout, TimeoutException.class),
+                Arguments.of(Disconnect, ConnectionException.class)
+                // can't test default branch in switch-case
+                );
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("getProtobufErrorsToJavaClientErrorsMapping")
+    @SneakyThrows
+    public void callback_dispatcher_handles_response_with_request_error(
+            RequestErrorType errorType, Class<? extends RedisException> exceptionType) {
+        var callbackDispatcher = new CallbackDispatcher();
+        var channelHandler = new TestChannelHandler(callbackDispatcher);
+        var commandManager = new CommandManager(channelHandler);
+
+        var future = commandManager.submitNewCommand(createDummyCommand(), r -> null);
+        var response =
+                Response.newBuilder()
+                        .setCallbackIdx(0)
+                        .setRequestError(RequestError.newBuilder().setType(errorType).setMessage("TEST"))
+                        .build();
+        callbackDispatcher.completeRequest(response);
+
+        var exception = assertThrows(ExecutionException.class, future::get);
+        // a RedisException thrown from CallbackDispatcher::completeRequest and then
+        // rethrown by CommandManager::exceptionHandler
+        assertEquals(exceptionType, exception.getCause().getClass());
+        assertEquals("TEST", exception.getCause().getMessage());
+        // check the channel
+        assertFalse(channelHandler.wasClosed);
+    }
+
+    @Test
+    @SneakyThrows
+    public void callback_dispatcher_handles_response_without_error_but_with_incorrect_callback_id() {
```

Comment:
Renamed in 0dbec7b9 + added a comment

Reply:
The fully correct description is too long, because callback dispatcher does not close the channel, it forces connection manager and command manager to do that.